### DISCUSSION TITLE
Using Pydantic for Configurations

### DISCUSSION THREAD

**Discussion by jmdelahanty** [wehr-lab/autopilot#159] [General]
Hello everyone!

I'm taking a break from trying to understand synaptic physiology by messing around with Pydantic. It'd be great to have class objects I manipulate instead of indexing different dictionaries like I currently do. [Here's](https://bruker-control.readthedocs.io/en/latest/configurations/configs.html#example-project-config-behavior) an example of a current configuration file. I think it'd be ideal to use Autopilot for these things and extend it so it can interface with Bruker microscopes, but in the meantime I'm just trying to learn pydantic and classes more generally. I figured I'd post here to see what people thought.

I'm wondering if I should change how my configuration files are formatted. Maybe something like this:

```
{"PARAMS": {
    "totalNumberOfTrials": 60,
    "startingReward": 3,
    "maxSequentialReward": null,
    "percentPunish": 0.25,
    "maxSequentialPunish": 3,
    "USDeliveryTime_Sucrose": 200,
    "USDeliveryTime_Air": 100,
    "USConsumptionTime_Sucrose": 3000,
    "vacuum" : true,
    "yoked": false,
    "dropped_frames": [],
    "weight_check": true,
    "ITIConfig": {
        "ITIJitter": true,
        "baseITI": null,
        "minITI": 15,
        "maxITI": 30
    },
    "toneConfig": {
        "punishTone": 2000,
        "rewardTone": 9000,
        "toneJitter": false,
        "baseTone": 3,
        "minTone": null,
        "maxTone": null
    },
    "stimConfig": {
        "stim": true,
        "shutterOnly": false,
        "stimFrequency": 20,
        "stimPulseTime": 5,
        "stimLambda": 590,
        "stimDeliveryTime_PreCS": 1000,
        "stimDeliveryTime_Total": 5000,
        "stimStartPosition": 20,
        "numStimReward": 5,
        "numStimPunish": 5,
        "numStimAlone": 3
    },
    "catchConfig": {
        "catchTrials": true,
        "numCatchReward": 3,
        "numCatchPunish": 2,
        "catchOffset" : 0.25
    },
    "stimulusArrays": {
        "trialArray": [],
        "ITIArray": [],
        "toneArray": [],
        "LEDArray": []
    },
    "zstackConfig": {
        "zstack": true,
        "stack_number": 3,
        "zdelta": 25.0,
        "zstep": 1.0}
    }
}
```

Or maybe something like this:

```
{"PARAMS": {
    "totalNumberOfTrials": 60,
    "startingReward": 3,
    "maxSequentialReward": null,
    "percentPunish": 0.25,
    "maxSequentialPunish": 3,
    "USDeliveryTime_Sucrose": 200,
    "USDeliveryTime_Air": 100,
    "USConsumptionTime_Sucrose": 3000,
    "vacuum" : true,
    "yoked": false,
    "dropped_frames": [],
    "weight_check": true,
    "ITIJitter": true,
    "baseITI": null,
    "minITI": 15,
    "maxITI": 30,
    "punishTone": 2000,
    "rewardTone": 9000,
    "toneJitter": false,
    "baseTone": 3,
    "minTone": null,
    "maxTone": null,
    "stim": true,
    "shutterOnly": false,
    "stimFrequency": 20,
    "stimPulseTime": 5,
    "stimLambda": 590,
    "stimDeliveryTime_PreCS": 1000,
    "stimDeliveryTime_Total": 5000,
    "stimStartPosition": 20,
    "numStimReward": 5,
    "numStimPunish": 5,
    "numStimAlone": 3,
    "catchTrials": true,
    "numCatchReward": 3,
    "numCatchPunish": 2,
    "catchOffset" : 0.25,
    "trialArray": [],
    "ITIArray": [],
    "toneArray": [],
    "LEDArray": [],
    "zstack": true,
    "stack_number": 3,
    "zdelta": 25.0,
    "zstep": 1.0
    }
}
```

Pydantic can make some nice subclasses for you automatically, and the first reformat looks like this on the python side:

```
class ITIConfig(BaseModel):
    iti_jitter: bool = Field(..., alias='ITIJitter')
    base_iti: Any = Field(..., alias='baseITI')
    min_iti: int = Field(..., alias='minITI')
    max_iti: int = Field(..., alias='maxITI')


class ToneConfig(BaseModel):
    punish_tone: int = Field(..., alias='punishTone')
    reward_tone: int = Field(..., alias='rewardTone')
    tone_jitter: bool = Field(..., alias='toneJitter')
    base_tone: int = Field(..., alias='baseTone')
    min_tone: Any = Field(..., alias='minTone')
    max_tone: Any = Field(..., alias='maxTone')


class StimConfig(BaseModel):
    stim: bool
    shutter_only: bool = Field(..., alias='shutterOnly')
    stim_frequency: int = Field(..., alias='stimFrequency')
    stim_pulse_time: int = Field(..., alias='stimPulseTime')
    stim_lambda: int = Field(..., alias='stimLambda')
    stim_delivery_time_pre_cs: int = Field(..., alias='stimDeliveryTime_PreCS')
    stim_delivery_time_total: int = Field(..., alias='stimDeliveryTime_Total')
    stim_start_position: int = Field(..., alias='stimStartPosition')
    num_stim_reward: int = Field(..., alias='numStimReward')
    num_stim_punish: int = Field(..., alias='numStimPunish')
    num_stim_alone: int = Field(..., alias='numStimAlone')


class CatchConfig(BaseModel):
    catch_trials: bool = Field(..., alias='catchTrials')
    num_catch_reward: int = Field(..., alias='numCatchReward')
    num_catch_punish: int = Field(..., alias='numCatchPunish')
    catch_offset: float = Field(..., alias='catchOffset')


class StimulusArrays(BaseModel):
    trial_array: List = Field(..., alias='trialArray')
    iti_array: List = Field(..., alias='ITIArray')
    tone_array: List = Field(..., alias='toneArray')
    led_array: List = Field(..., alias='LEDArray')


class ZstackConfig(BaseModel):
    zstack: bool
    stack_number: int
    zdelta: float
    zstep: float


class Params(BaseModel):
    total_number_of_trials: int = Field(..., alias='totalNumberOfTrials')
    starting_reward: int = Field(..., alias='startingReward')
    max_sequential_reward: Any = Field(..., alias='maxSequentialReward')
    percent_punish: float = Field(..., alias='percentPunish')
    max_sequential_punish: int = Field(..., alias='maxSequentialPunish')
    us_delivery_time_sucrose: int = Field(..., alias='USDeliveryTime_Sucrose')
    us_delivery_time_air: int = Field(..., alias='USDeliveryTime_Air')
    us_consumption_time_sucrose: int = Field(..., alias='USConsumptionTime_Sucrose')
    vacuum: bool
    yoked: bool
    dropped_frames: List
    weight_check: bool
    iti_config: ITIConfig = Field(..., alias='ITIConfig')
    tone_config: ToneConfig = Field(..., alias='toneConfig')
    stim_config: StimConfig = Field(..., alias='stimConfig')
    catch_config: CatchConfig = Field(..., alias='catchConfig')
    stimulus_arrays: StimulusArrays = Field(..., alias='stimulusArrays')
    zstack_config: ZstackConfig = Field(..., alias='zstackConfig')
```

If I do it the second way, it looks like this:

```
class Params(BaseModel):
    total_number_of_trials: int = Field(..., alias='totalNumberOfTrials')
    starting_reward: int = Field(..., alias='startingReward')
    max_sequential_reward: Any = Field(..., alias='maxSequentialReward')
    percent_punish: float = Field(..., alias='percentPunish')
    max_sequential_punish: int = Field(..., alias='maxSequentialPunish')
    us_delivery_time__sucrose: int = Field(..., alias='USDeliveryTime_Sucrose')
    us_delivery_time__air: int = Field(..., alias='USDeliveryTime_Air')
    us_consumption_time__sucrose: int = Field(..., alias='USConsumptionTime_Sucrose')
    vacuum: bool
    yoked: bool
    dropped_frames: List
    weight_check: bool
    iti_jitter: bool = Field(..., alias='ITIJitter')
    base_iti: Any = Field(..., alias='baseITI')
    min_iti: int = Field(..., alias='minITI')
    max_iti: int = Field(..., alias='maxITI')
    punish_tone: int = Field(..., alias='punishTone')
    reward_tone: int = Field(..., alias='rewardTone')
    tone_jitter: bool = Field(..., alias='toneJitter')
    base_tone: int = Field(..., alias='baseTone')
    min_tone: Any = Field(..., alias='minTone')
    max_tone: Any = Field(..., alias='maxTone')
    stim: bool
    shutter_only: bool = Field(..., alias='shutterOnly')
    stim_frequency: int = Field(..., alias='stimFrequency')
    stim_pulse_time: int = Field(..., alias='stimPulseTime')
    stim_lambda: int = Field(..., alias='stimLambda')
    stim_delivery_time__pre_cs: int = Field(..., alias='stimDeliveryTime_PreCS')
    stim_delivery_time__total: int = Field(..., alias='stimDeliveryTime_Total')
    stim_start_position: int = Field(..., alias='stimStartPosition')
    num_stim_reward: int = Field(..., alias='numStimReward')
    num_stim_punish: int = Field(..., alias='numStimPunish')
    num_stim_alone: int = Field(..., alias='numStimAlone')
    catch_trials: bool = Field(..., alias='catchTrials')
    num_catch_reward: int = Field(..., alias='numCatchReward')
    num_catch_punish: int = Field(..., alias='numCatchPunish')
    catch_offset: float = Field(..., alias='catchOffset')
    trial_array: List = Field(..., alias='trialArray')
    iti_array: List = Field(..., alias='ITIArray')
    tone_array: List = Field(..., alias='toneArray')
    led_array: List = Field(..., alias='LEDArray')
    zstack: bool
    stack_number: int
    zdelta: float
    zstep: float
```

I think both are better ways of accessing data in the program, but the thing I'm unsure about is how to make new classes that are composed of just some pieces of these fields. For example, at the moment with the Arduino, these fields are required to set the board up to do the correct stimuli:

```
class ArduinoConfig(Params):
    """
    Class for metadata relevant to Arduino Configuration that is sent via serial_transfer_utils.
    """
    total_number_of_trials: int = Params.total_number_of_trials
    punish_tone: int = Params.punish_tone
    reward_tone: int = Params.reward_tone
    us_delivery_time_sucrose: Params.us_delivery_time_sucrose
    us_delivery_time_air: int = Params.us_delivery_time_air
    us_consumption_time_sucrose: int = Params.us_consumption_time_sucrose
```

This takes some pieces of one class while taking different pieces of another. I feel like this means I'm just doing all of this badly. Are both of these organizations garbage? Does it make sense to make a new class that's just composed of different pieces of other classes?

Curious what you all think...

**Comment by sneakers-the-rat** [comment]
With these kinds of types, the type model is part of the art. 

One way of thinking about it is hierarchically: you have a certain number of basic types that only consist of builtins/etc., and then higher-order types that integrate them. So then in that framing the problem of needing to subset some fields of a type is a problem of not having composed it of fine enough components, aggregating too soon, etc. For example your 'top level' params class could resolve that in at least two ways: one is that you make that group of params as arduino params, and then include it in the top-level class. Another is that both the top-level params object and the arduino params object share some "lower level" parameter types. 

There's a distinction between three things here, the first two of which are familiar from regular python:
- class, or the type description, which describes which attributes are present in any of the instances
- instances, or specific applications of a class that fill the attributes
- types, descriptions of what kinds of values a class accepts for its instances. 

so the first mode of nesting types is strictly preferable, as being able to compose individual attributes into recognizable groups is the whole idea. 

In the last example, it looks like you're mixing up the notion of type and default, and mixing up the "direction" of inheritance. It seems like you want to make a subset of the `Params` class, but inheriting from it will give you all the parameters from the `Params` class *plus* any new ones you create/override. Declaring `punish_tone: int = Params.punish_tone` means that `punish_tone` is an integer whose default is the class attribute of `Params.punish_tone`, which will be the `Field` description rather than the value (as there is no default set, just `...` as the abstraction of not having a value in pydantic). 

I think it might be useful to check out the docs for the [typing](https://docs.python.org/3/library/typing.html) module as well, as there are plenty of places where it looks like you'd probably want to declare, for example, a list of integers like `List[int]`instead of just `List`.

**Reply by jmdelahanty** [reply]
> so the first mode of nesting types is strictly preferable, as being able to compose individual attributes into recognizable groups is the whole idea.

If I'm understanding this correctly, this means that the second version which has _everything_ in, so no specific fields like `toneConfig`, is a better way to start so I can then compose classes from those individual pieces more flexibly. If that's the appropriate way, then I'm not sure how to build classes in the correct direction, where I have the base `Params` class and then build new classes on top of it. Is that something I can do properly in Pydantic? Should I have specific fields in my json that are organized like that? Sorry for all these super basic questions.

> you make that group of params as arduino params, and then include it in the top-level class.

I was thinking of trying to do this, but that seemed to mean that I had to repeat data in the configuration file which seems like something I shouldn't do either...

> In the last example, it looks like you're mixing up the notion of type and default, and mixing up the "direction" of inheritance. It seems like you want to make a subset of the Params class, but inheriting from it will give you all the parameters from the Params class plus any new ones you create/override.

I didn't realize this at all! Thank you for explaining that.

> I think it might be useful to check out the docs for the [typing](https://docs.python.org/3/library/typing.html) module as well

Thank you for this too! It's really cool you can offer type-hints and validation for classes.

**Reply by jmdelahanty** [reply]
Update: I think I'm getting the hang of this now! One of my friends explained some things to me about inheritance and I think I have a better grip on things. The key for my understanding was that you should start from the most re-usable part of the object and build more complex objects as needed. In this case, almost literally every object needs to have the `totalNumberOfTrials` available so I can generate arrays for the Arduino. When I finally move away from the Arduino and adopt Autopilot's style for running trials, I guess much of this won't be used anymore. But it's a great learning experience. The next thing to incorporate is the metadata of a given subject so I can personalize the imaging session. I guess I can have a second main model for that? What do you think?

Here's my new proposed configuration file:

```
{"PARAMS": {
    "totalNumberOfTrials": 60,
    "trialSettings": {
        "percentPunish": 0.50,
        "maxSequentialReward": null,
        "maxSequentialPunish": 3,
        "startingReward": 3,
        "catchTrials": true,
        "numCatchReward": 3,
        "numCatchPunish": 2,
        "catchOffset": 0.25,
        "stimSettings": {
            "stim": true,
            "shutterOnly": false,
            "stimReward": 3,
            "stimPunish": 3,
            "stimAlone": 3,
            "stimStartPosition": 20,
            "LED": {
                "stimDeliveryTimePrecs": 1000,
                "stimDeliveryTimeTotal": 5000,
                "stimFrequency": 20,
                "stimPulseTime": 5,
                "stimLambda": 590
            } 
        }
    },
    "itiSettings": {
        "itiJitter": false,
        "baseITI": null,
        "minITI": 15,
        "maxITI": 30
    },
    "toneSettings": {
        "toneJitter": false,
        "baseTone": null,
        "minITI": 15,
        "maxITI": 30
    },
    "arduinoMetadata": {
        "punishTone": 2000,
        "rewardTone": 9000,
        "USDeliveryTimeSucrose": 100,
        "USDeliveryTimeAir": 50,
        "USConsumptionTimeSucrose": 500
    },
    "behaviorArrays": {
        "trialArray": [],
        "itiArray": [],
        "toneArray": [],
        "LEDArray": []
    },
    "zstackMetadata": {
        "zstack": true,
        "stackNumber": 3,
        "zDelta": 25.0,
        "zStep": 1.0
    }
  }
}
```

And here's my proposed Pydantic model:
```
class Led(BaseModel):
    stim_delivery_time_precs: int = Field(..., alias='stimDeliveryTimePrecs')
    stim_delivery_time_total: int = Field(..., alias='stimDeliveryTimeTotal')
    stim_frequency: int = Field(..., alias='stimFrequency')
    stim_pulse_time: int = Field(..., alias='stimPulseTime')
    stim_lambda: int = Field(..., alias='stimLambda')


class StimSettings(BaseModel):
    stim: bool
    shutter_only: bool = Field(..., alias='shutterOnly')
    stim_reward: int = Field(..., alias='stimReward')
    stim_punish: int = Field(..., alias='stimPunish')
    stim_alone: int = Field(..., alias='stimAlone')
    stim_start_position: int = Field(..., alias='stimStartPosition')
    led: Led = Field(..., alias='LED')


class TrialSettings(BaseModel):
    percent_punish: float = Field(..., alias='percentPunish')
    max_sequential_reward: Any = Field(..., alias='maxSequentialReward')
    max_sequential_punish: int = Field(..., alias='maxSequentialPunish')
    starting_reward: int = Field(..., alias='startingReward')
    catch_trials: bool = Field(..., alias='catchTrials')
    num_catch_reward: int = Field(..., alias='numCatchReward')
    num_catch_punish: int = Field(..., alias='numCatchPunish')
    catch_offset: float = Field(..., alias='catchOffset')
    stim_settings: StimSettings = Field(..., alias='stimSettings')


class ItiSettings(BaseModel):
    iti_jitter: bool = Field(..., alias='itiJitter')
    base_iti: Any = Field(..., alias='baseITI')
    min_iti: int = Field(..., alias='minITI')
    max_iti: int = Field(..., alias='maxITI')


class ToneSettings(BaseModel):
    tone_jitter: bool = Field(..., alias='toneJitter')
    base_tone: Any = Field(..., alias='baseTone')
    min_iti: int = Field(..., alias='minITI')
    max_iti: int = Field(..., alias='maxITI')


class ArduinoMetadata(BaseModel):
    punish_tone: int = Field(..., alias='punishTone')
    reward_tone: int = Field(..., alias='rewardTone')
    us_delivery_time_sucrose: int = Field(..., alias='USDeliveryTimeSucrose')
    us_delivery_time_air: int = Field(..., alias='USDeliveryTimeAir')
    us_consumption_time_sucrose: int = Field(..., alias='USConsumptionTimeSucrose')


class BehaviorArrays(BaseModel):
    trial_array: List[int] = Field(..., alias='trialArray')
    iti_array: List[int] = Field(..., alias='itiArray')
    tone_array: List[int] = Field(..., alias='toneArray')
    led_array: List[int] = Field(..., alias='LEDArray')


class ZstackMetadata(BaseModel):
    zstack: bool
    stack_number: int = Field(..., alias='stackNumber')
    z_delta: float = Field(..., alias='zDelta')
    z_step: float = Field(..., alias='zStep')


class Params(BaseModel):
    total_number_of_trials: int = Field(..., alias='totalNumberOfTrials')
    trial_settings: TrialSettings = Field(..., alias='trialSettings')
    iti_settings: ItiSettings = Field(..., alias='itiSettings')
    tone_settings: ToneSettings = Field(..., alias='toneSettings')
    arduino_metadata: ArduinoMetadata = Field(..., alias='arduinoMetadata')
    behavior_arrays: BehaviorArrays = Field(..., alias='behaviorArrays')
    zstack_metadata: ZstackMetadata = Field(..., alias='zstackMetadata')


class Model(BaseModel):
    params: Optional[Params] = Field(None, alias='PARAMS')

```

I still need to rewrite the `trial_utils` module to handle these classes and make some methods for populating the arrays, but this feels like progress. 

Edit: Reintroduce `List[int]` type hint.

**Reply by sneakers-the-rat** [reply]
Nice, that looks pretty good to me. Also check out the alias generator so you don't need to do `Field(...,alias='camelCase')` for every field: https://pydantic-docs.helpmanual.io/usage/model_config/#alias-generator

**Reply by jmdelahanty** [reply]
Thanks Jonny! I'll definitely use the alias generator instead of writing it out for every field, definitely seems much nicer.

**Reply by jmdelahanty** [reply]
Hey Jonny! I'm making some good progress with refactoring my code to use classes but am pretty stuck with trying to alter my classes properly.

At the moment, I have lots of functions that can now take in my class objects and perform operations on the values my classes contain, but I feel like I should be turning them into class methods instead.

For example, here's a function I have for making an array of ITIs:
```
def gen_static_ITIArray(config):
    # Get the base ITI to use for the session and multiply by 1000 to convert
    # to milliseconds.
    iti_duration = config.iti_settings.base_iti*1000

    # Build ITIArray into a list of values
    config.behavior_arrays.iti_array = [iti_duration] * config.total_number_of_trials

    return config
```

I've tried turning this into a class method and a static method in various places in my model, but can't seem to get it right. Is the way I'm doing it a good way or should I keep hacking away at making these into methods I invoke on the model?

If I do end up making these into class methods, I'm not sure how to invoke `if`/`else` statements for them also. For example, if `iti_jitter` is `True`, then I want to call `gen_jitter_ITIArray` and if it's `False` I want to call `gen_static_ITIArray`. It seems to be much easier to just have these little functions update stuff as I go along, but I feel like it's bad practice.

Any advice?

**Reply by sneakers-the-rat** [reply]
sounds like you're looking for `@property`

```python
class SomeModel(BaseModel):

    stuff: str = "whatever"
    
    @property
    def ITIArray(self) -> np.ndarray:
        # whatever calculation
        iti_array = result_of_calculation
        
        if self.jitter:
            #do the jitter
        return iti_array 
```

**Reply by sneakers-the-rat** [reply]
definitely discourage using functions to update stuff in the model instances -- it sort of defeats the purpose. the whole point of having these explicit data models is that you can verifiably rely on certain values existing and having a certain type. if you need to call a bunch of functions on a model after you make it then you lose that! if you have some attributes/arrays/values that are derived from the parameters, use properties or instance methods. Class methods are for when the class as a whole needs to be able to do something. a common pattern is to do something like this

```python
class MyClass:

    @classmethod
    def from_string(cls, input:str) -> MyClass:
         # do something to create the class from a string
         return cls(whatever)
```

**Reply by sneakers-the-rat** [reply]
also, definitely do type hinting. another one of those "that's the whole point" kind of things.

**Reply by jmdelahanty** [reply]
Thank you so much for these tips! That makes sense that you'd want to be doing it all through the classes since it is literally the point of them lol. Is there a way to provide attributes from higher up in the model? For example, the ITI array I need to make needs to have the number of elements inside equal to the total number of trials. I'm not sure how to feed in the `total_number_of_trials` attribute of my `Params` class. I thought my class would have inherited what was at the top level, but I think I might still have how inheritance works backwards...

**Reply by sneakers-the-rat** [reply]
can i see a minimal example, in pseudocode if needed, of the model you're thinking of and what you want it to do? brain fried and can't quite parse this/the above models.

when you can't figure out if something is working as you expect, i like to make toy examples. so to see how inheritance is different than being a class attribute something like

```python
class Child:
   b = 'b'
   
class Parent:
    a = 'a'
    child = Child()
    
# does my child have access to the parent value?
>>> assert hasattr(Child(), 'a')
```

vs.

```python
class Parent:
    a = 'a'
    
class Child(Parent):
    b = 'b'

# does my child have access to the parent value?
assert hasattr(Child(), 'a')

```

**Reply by jmdelahanty** [reply]
I changed where the property goes and it seems to work! I had the inheritance backwards again.

The organization of it basically is (pseudocode for model then real property):

```
class ITISettings(BaseModel):
     settings: str
    ....

class Model(BaseModel):
     iti_settings = ITISettings
     ....

     @property
     def ITIArray(self) -> List:
        
        if self.params.iti_settings.iti_jitter:

            rng = default_rng()

            iti_array = rng.uniform(
                low=self.params.iti_settings.min_iti,
                high=self.params.iti_settings.max_iti,
                size=self.params.total_number_of_trials
            )

        else:

            iti_array = [self.params.iti_settings.base_iti] * self.params.total_number_of_trials
        
        iti_array =  np.round(iti_array).astype(int).tolist()

        return iti_array
```

But then I have all these other properties at the top level of the structure, meaning I will have a bunch more properties tags beneath my top level class. Is it bad to have them at that level design wise? I also feel like having to do `self.params.subclass.variable` every time is a little messy too...

EDIT:

For better or worse, I know can generate all my arrays using classes and I access them all as properties! Woohoo! I still have to do `self.params.subclass.variable` for each of them, but it actually feels like it's surprisingly more readable maybe because of how explicit it is despite it being a lot of text...
